### PR TITLE
pydantic config (#5)

### PR DESCRIPTION
**Pull request by clementlrd**
I created the first blocks of our settings interface using pydantic. Please let me know what you think about the structure.

I also added settings that can be changed using a .env file, this can be used to change the config path (because we need to know where to find the config file before using it), this can also be useful for tests if we want to change variables for testing purposes.

Related Issue: #5

### CODE REVIEW

**Comment by toniodo** [comment]
Thank you for your work. I don't really understand what's the difference between the config and settings class. Maybe one has accessed to only read-only fields and not the other. Because they seem to act in the same way. The only difference could be the file used to look at the fields (.env for one and .json file for the other one). 
And also, I don't think we need the import of annotations for now.

**Comment by clementlrd** [comment]
Thank you for your comment. You're right the annotation import seems to have nothing to do here.

BaseSettings come from [pydantic-settings](https://docs.pydantic.dev/dev/concepts/pydantic_settings/) module, an extension of pydantic that load out of the box a .env file, which can be useful for our case, at least for the CONFIG_FILE field if we want to update it; but also later if we add other fields that can be set in a .env file (as for testing for example: we can have a .env.test file that loads specific fields for testing purposes).

On the other hand, BaseModel is the classic [pydantic](https://docs.pydantic.dev/dev/) object. As a result, it provides data validation for the provided fields (and thus crash in that case if there is a problem in the configuration). The Config object will be the class where we store the configurable fields.

As mentioned with the TODO, the loading of the config as not been written yet, but it is just the loading of a json file, then we give all the fields to the init function for creating the Config object.

They are both Singleton instance because they are global settings and do not need to be recreated every time. Adding them in the __init__.py file of the module ensure that tey are loaded before anything else happen.

Fell free to ask if you have any questions.

**Comment by toniodo** [comment]
Oh, you're right I didn't notice this little difference that changes everything of course, I understand now why you made the distinction using those classes. And, thank you for the reminder concerning pydantic tools.

**Comment by clementlrd** [comment]
`.env` file shouln't be removed from the gitignore file, at least only whithelist a `.env.example`

**Comment by Seb-sti1** [comment]
I don't really see the point of having a `.env`... For testing purpose we can use `core.json` right? 

> which can be useful for our case, at least for the CONFIG_FILE field

Regarding this, it don't see why you would need to change the `CONFIG_FILE` but more importantly, why i don't see how it is convenient to change it using the .env in the source of the app for the user. And if we change the `CONFIG_FILE` then it would break every user's setup on update, which would mean it would require some king of migration...


Also if the user is supposed to edit it, I think yaml files are more user-friendly.

**Review by Seb-sti1 — COMMENTED** [review]
I don't really see the point of having a `.env`... For testing purpose we can use `core.json` right? 

> which can be useful for our case, at least for the CONFIG_FILE field

Regarding this, it don't see why you would need to change the `CONFIG_FILE` but more importantly, why i don't see how it is convenient to change it using the .env in the source of the app for the user. And if we change the `CONFIG_FILE` then it would break every user's setup on update, which would mean it would require some king of migration...


Also if the user is supposed to edit it, I think yaml files are more user-friendly.

**Comment by clementlrd** [comment]
I haven't yet thought of how to make it user-configurable. For sure people can't edit the .env file themselves , it has to be done from a command in the command line.

For me, a simple example where we would like to have different location for the config file is between production and development. I don't want to do a backup of my config every time I want to work on the project as a developer. A .env file can be useful for that. But in #22, I also used the .env file to enable/disable the traceback to have a more elegent display in production. This may change in the future but at least loading environment variables can be useful in many ways.

**Comment by Seb-sti1** [comment]
> I haven't yet thought of how to make it user-configurable. For sure people can't edit the .env file themselves , it has to be done from a command in the command line.
> 
> For me, a simple example where we would like to have different location for the config file is between production and development. I don't want to do a backup of my config every time I want to work on the project as a developer. A .env file can be useful for that. But in #22, I also used the .env file to enable/disable the traceback to have a more elegent display in production. This may change in the future but at least loading environment variables can be useful in many ways.

Okay i see. What about the yaml instead of json?

**Comment by clementlrd** [comment]
> What about the yaml instead of json?

For the config file ? This is you that wanted a json instead of a yaml file x) But at the end it is also easy to use with pydantic. YAML file can be nearly as easily parsed I think with the pyyaml library (and the `safe_load` function) so it's up to you. We can also choose to support both and it depends on the config path... It's not a lot of code to do that...

File: projecthub/config.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+class _Settings(BaseSettings):
+    """Automatically loads the settings from a `.env` file.
+
+    This is designed to be a Singleton instance. 
+    The settings are loaded once and can then be accessed using the Setting instance.
+    """
+
+    model_config = SettingsConfigDict(env_file=DOTENV_FILE)
+
+    # The Path does not necessary exists, use FilePath otherwise
```
Comment (toniodo) [inline projecthub/config.py:26]:
```suggestion
    # The Path does not necessarily exists, use FilePath otherwise
```
Typo

**Comment by toniodo** [comment]
To summarize the main idea : the `.env` file is the configuration of the app, and is edited through the CLI, and the `.json` is the configuration concerning the projects of the user. Do you agree with that?

**Review by toniodo — DISMISSED** [review]

**Comment by Seb-sti1** [comment]
> > What about the yaml instead of json?
> 
> For the config file ? This is you that wanted a json instead of a yaml file x) But at the end it is also easy to use with pydantic. YAML file can be nearly as easily parsed I think with the pyyaml library (and the `safe_load` function) so it's up to you. We can also choose to support both and it depends on the config path... It's not a lot of code to do that 😉

What i said is that i prefer json because i feel like it's more reliable and i usually don't use yaml... but i think yaml is more user-friendly. If supporting both is easy, then we can do that too

**Comment by clementlrd** [comment]
@Seb-sti1 I will add yaml support and set it as the default type for the config file.

**Comment by clementlrd** [comment]
**TL;DR:** There are two objects:
- The Config object for the config file (`config.[json|yaml]`)
- The Settings object for environment variables (`.env.*`)
___

To explain in more depth for @toniodo, there is two distinct things here. Maybe I shoud have split the PR in two.

- The `Config` object mirrors of the fields specified in the config file. This is where the field are loaded with the values specified and if they are not specified, we have to provide the default values (this is done through the `default` argument of the `Field` object in pydantic).

Now when we think about the config file location, we can provide a default value, but it would also be welcome to be able to specify a custom path. But we can't put the path of the config file in the config file because when we load the config file we need to know where to look. On the other hand, we discussed with @Seb-sti1 the potential usefulness of loading environment variables (from .env for example) and especially for the loacation of the config file because it may be useful to have different path depending on production/development mode. 
- So I also added the `Settings` object that automatically loads data from the .env file but also directly from the loaded environment variables. So it is also possible to do something like this:
`CONFIG_PATH="~/projecthub.json" projecthub --hello`.

But environment variables files SHOULD be used rarely and only be fields that are environment dependent. The configurable fields one want to customize for the app all go into the config file as they should.

**Review by Seb-sti1 — APPROVED** [review]

**Review by toniodo — APPROVED** [review]

**Review by toniodo — APPROVED** [review]
